src/librustc_mir/borrow_check/nll/constraint_generation.rs

```diff
@@ -108,15 +108,24 @@ impl<'cx, 'cg, 'gcx, 'tcx> ConstraintGeneration<'cx, 'cg, 'gcx, 'tcx> {
                     });
 
                     let mpi = self.move_data.rev_lookup.find_local(live_local);
-                    if let Some(initialized_child) = self.flow_inits.has_any_child_of(mpi) {
+                    let mut any = false;
+                    for initialized_child in self.flow_inits.initialized_children_of(mpi) {
+                        any = true;
                         debug!("add_liveness_constraints: mpi={:?} has initialized child {:?}",
                                self.move_data.move_paths[mpi],
                                self.move_data.move_paths[initialized_child]);
+
+                        let initialized_child_ty = self.move_data
+                            .move_paths[initialized_child]
+                            .lvalue
+                            .ty(self.mir, self.infcx.tcx)
+                            .to_ty(self.infcx.tcx);
+                        self.add_drop_live_constraint(initialized_child_ty, location);
+                    }
+
+                    if any {
```

Comment:
I'm not sure if this was still necessary or not. Also, the test is still not working as expected. For one, it's still in /ui. I thought I had moved it to /compile-fail, but I might have lost that work when rebasing.